firedrake/adjoint_utils/variational_solver.py

```diff
@@ -133,5 +133,6 @@ def _ad_problem_clone(self, problem, dependencies):
                                            F_replace_map[problem.u_restrict],
                                            bcs=problem.bcs,
                                            J=replace(problem.J, J_replace_map))
+        nlvp._constant_jacobian = problem._constant_jacobian
```

Comment:
I'm not really happy with this. Should a `NonlinearVariationalProblem` set the `_constant_jacobian` attribute?

Reply:
Why not? This is a case where the NonlinearVariationalProblem is known to actually be linear.

Reply:
This is setting an attribute that a `NonlinearVariationalProblem` (base class) doesn't otherwise have.

Reply:
so how does this work in plain firedrake (ie. not in the adjoint)?

Reply:
A `LinearVariationalProblem` does have this attribute. In the adjoint the `LinearVariationalProblem` is always replaced internally with a `NonlinearVariationalProblem`.

Reply:
Oh, this sounds like a minor subclassing design error. Probably nvlp should always have it and it should be false. I wouldn't be too concerned.